RockLib.Messaging.SNS/DependencyInjection/SNSExtensions.cs

```diff
@@ -26,7 +26,7 @@ public static class SNSExtensions
             return services.AddSender(name, CreateSNSSender, configureOptions, reloadOnChange);
 
             ISender CreateSNSSender(SNSSenderOptions options, IServiceProvider serviceProvider) =>
-                new SNSSender(name, options.TopicArn, options.Region);
+                new SNSSender(name, options.TopicArn, options.Region, options.MessageGroupId);
```

Comment:
Wouldn't this cause a problem for singleton registered senders? Not sure if this is norm or not, but gave me pause thinking about possible use cases.

Reply:
In this case it is registered as a singleton, it just uses the default ServiceLifetime in the RockLib.Messaging.DependencyInjection.MessagingExtensions. 

What kind of problem are you thinking though?

Reply:
@NeilSchiefer - My assumption is that with the current implementation the sender will always be using the same `MessageGroupID` on every request, which could be the intended result. 

My thought was this was to group by something more like a SessionId/RequestId (or similar) that is related to the request. Having this in the `ctor` for the sender doesn't suggest that kind of behavior, to me. If the intent is more like what I'm expecting then it would make more sense to have the `MessageGroupId` as a part of the `SendAsync` not the configuration of the sender. Being on the sender I don't see working for the kind of grouping I was assuming, at least with the possibility of concurrency.

Reply:
Ah, I see.  Maybe we should remove MessageGroupId as a property and possibly just pass it to the SendAsync explicitly.  However, that may be a lot of work due to the ISender interface.  An alternative would be to just use the headers.  Also we have a request to add MessageDeduplicationId, so maybe we could re-work this and add that as well right away.

I'd like to get the rest of the maintainers take on this after they get back from break.

Reply:
Hi folks, I'm chiming in to explain my reasoning for this design. More accurately, my lack of designing because I just copied the design and API of the SQS sender 🙂 

https://github.com/RockLib/RockLib.Messaging/blob/4e99e7062e0fb908df7eb960a33a9a36414c46c6/RockLib.Messaging.SQS/SQSSender.cs#L48

Reply:
This design (and the `SQSSender`) is misleading to the user and in some ways defeats the purpose of the group ID. The group ID should be on a per message basis. I think this should be reworked to have the group id incorporated with the `Send` method either as an argument or part of the `SenderMessage`. 

Where this makes sense is in the case of a default message group ID, but then that adds another level of complexity, managing multiple ways of providing the group ID.

Reply:
@aatchison14 reached out to me this morning about this PR, and getting it into good shape in terms of moving the message group ID from the sender to the message.

I'll refresh myself on this PR today and see about quickly making those changes.

Reply:
Ok so if, as @pjwilliams2 suggests, we move the group ID parameter from the sender to the sending, that will be  a breaking change to the SQS Sender API.

One approach to address that is to introduce the group ID parameter in an overload of the `ISender.SendAsync` method, and have that argument override the group ID that may or may not have been set on the sender object. Then, put `[Obsolete]` attributes on the Sender-level parameter. Finally, remove the Sender-level parameter.

Reply:
Looking at this again this morning with fresh eyes, I realized I missed the other option that @pjwilliams2 outlined: Putting the message group ID on the `SenderMessage` object. That is quite possibly a better option.